routers/private/hook.go

```diff
@@ -45,206 +34,258 @@ func HookPreReceive(ctx *macaron.Context) {
+				PusherID:     opts.UserID,
+				PusherName:   opts.UserName,
+				RepoUserName: ownerName,
+				RepoName:     repoName,
 			})
-			return
 		}
 	}
 
-	if newCommitID != git.EmptySHA && strings.HasPrefix(refFullName, git.BranchPrefix) {
-		repo, err := models.GetRepositoryByOwnerAndName(ownerName, repoName)
-		if err != nil {
-			log.Error("Failed to get repository: %s/%s Error: %v", ownerName, repoName, err)
-			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
-				"err": fmt.Sprintf("Failed to get repository: %s/%s Error: %v", ownerName, repoName, err),
-			})
-			return
+	if err := repofiles.PushUpdates(repo, updates); err != nil {
+		log.Error("Failed to Update: %s/%s Total Updates: %d", ownerName, repoName, len(updates))
+		for i, update := range updates {
+			log.Error("Failed to Update: %s/%s Update: %d/%d: Branch: %s", ownerName, repoName, i, len(updates), update.Branch)
 		}
-		repo.OwnerName = ownerName
+		log.Error("Failed to Update: %s/%s Error: %v", ownerName, repoName, err)
 
-		pullRequestAllowed := repo.AllowsPulls()
-		if !pullRequestAllowed {
-			ctx.JSON(http.StatusOK, map[string]interface{}{
-				"message": false,
-			})
-			return
+		ctx.JSON(http.StatusInternalServerError, []private.HookPostReceiveResult{
+			{
+				Err: fmt.Sprintf("Failed to Update: %s/%s Error: %v", ownerName, repoName, err),
+			},
+		})
+		return
+	}
+
+	results := make([]private.HookPostReceiveResult, 0, len(opts.OldCommitIDs))
+
+	// We have to reload the repo in case its state is changed above
+	repo = nil
+	var baseRepo *models.Repository
+
+	for i := range opts.OldCommitIDs {
+		refFullName := opts.RefFullNames[i]
+		newCommitID := opts.NewCommitIDs[i]
+
+		branch := opts.RefFullNames[i]
+		if strings.HasPrefix(branch, git.BranchPrefix) {
+			branch = strings.TrimPrefix(branch, git.BranchPrefix)
+		} else {
+			branch = strings.TrimPrefix(branch, git.TagPrefix)
 		}
 
-		baseRepo := repo
-		if repo.IsFork {
-			if err := repo.GetBaseRepo(); err != nil {
-				log.Error("Failed to get Base Repository of Forked repository: %-v Error: %v", repo, err)
-				ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
-					"err": fmt.Sprintf("Failed to get Base Repository of Forked repository: %-v Error: %v", repo, err),
-				})
-				return
+		if newCommitID != git.EmptySHA && strings.HasPrefix(refFullName, git.BranchPrefix) {
+			if repo == nil {
+				var err error
+				repo, err = models.GetRepositoryByOwnerAndName(ownerName, repoName)
+				if err != nil {
+					log.Error("Failed to get repository: %s/%s Error: %v", ownerName, repoName, err)
+					ctx.JSON(http.StatusInternalServerError, []private.HookPostReceiveResult{
+						{
+							Err: fmt.Sprintf("Failed to get repository: %s/%s Error: %v", ownerName, repoName, err),
+						},
+					})
+					return
+				}
+				if repo.OwnerName == "" {
+					repo.OwnerName = ownerName
+				}
+
+				pullRequestAllowed := repo.AllowsPulls()
+				if !pullRequestAllowed {
```

Comment:
This variable doesn't seem to be required (`repo.AllowPulls()` can be invoked in the if).

Reply:
Done